File: lab8/3-concurrency/concurrency.cpp
#include <thread>
#include <mutex>
#include <vector>
#include <iostream>
#include <atomic>

typedef struct account {
    int balance;
    std::mutex balance_mtx;
} account_t;

/************************************** HELPER FUNCTION **************************************/

/* This function initializes a very large vector (5000 elts long) containing all the changes we'd like to 
   make to the account balance. 
*/
std::vector<int> make_large_vector() {
    std::vector<int> v;

    for (int i = 0; i < 5000; i++) {
        v.push_back(i - 300);
    }

    return v;
}

/************************************** THREAD FUNCTION **************************************/

/* This function will get passed into threads in order to modify an account with the requests. 
   Arguments: 
    - account_to_update: a pointer to the account whose balance we wish to modify
    - requests: a pointer to the vector containing all the requests
    - requests_mtx: a pointer to the mutex for the shared requests vector
    - requests_fulfilled: a pointer to the atomic boolean that dictates whether the threads can stop or not
*/
void update_account(account_t* account_to_update, std::vector<int>* requests, std::mutex* requests_mtx, std::atomic_bool* requests_fulfilled) {
    while (!*requests_fulfilled) {
        int current_request;
        requests_mtx->lock();
        
        if (requests->size() == 0) { 
            // Checks whether the requests are all fulfilled, and if so, set the boolean to true so other threads can stop
            *requests_fulfilled = true;
            requests_mtx->unlock();
            return;
        } else {
            // Otherwise, grab the first request
            current_request = requests->at(0);
            
            // Erase the request from the beginning of the vector and unlock the vector for other threads
            requests->erase(requests->begin());
            requests_mtx->unlock();

            // Add the amount in the request to the account's balance
            account_to_update->balance += current_request;
        }
    }   
}

/************************************** MAIN PROGRAM **************************************/

int main() {
    // Initialize an account with a balance
    account_t act;
    act.balance = 300;

    // Initialize vector with all the requests to add/subtract money from account
    std::vector<int> requests = make_large_vector();
    std::mutex requests_mtx;

    // atomic boolean that will get set to true when all requests have been fulfilled (so workers can stop working)
    std::atomic_bool all_requests_fulfilled = false;

    // Initialize worker pool and pass `update_account` function into threads
    std::vector<std::thread> workers;

    for (int i = 0; i < 6; i++) {
        workers.push_back(std::thread(update_account, &act, &requests, &requests_mtx, &all_requests_fulfilled));
    }
    
    // Wait for all threads to finish
    for (int i = 0; i < 6; i++) {
        workers.at(i).join();
    }

    // Print final balance
    std::cout << "Final balance = " << act.balance << std::endl;

    return 0;
}

File: lab8/3-concurrency/concurrency.rs
use std::sync::{Mutex, Arc};
use std::sync::atomic::{AtomicBool, Ordering};
use std::thread;
        
struct Account {
    balance: Mutex<i32>
}

/************************************** HELPER FUNCTION **************************************/

/* This function initializes a very large vector (5000 elts long) containing all the changes we'd like to 
   make to the account balance. 
*/
fn make_large_vector() -> Vec<i32> {
    let mut v = vec![];
    for i in 0..5000 {
        v.push(i - 300);
    }
    return v
}

/************************************** THREAD FUNCTION **************************************/

/* This function will get passed into threads in order to modify an account with the requests. 
   Arguments: 
    - account_to_update: a reference to the account whose balance we wish to modify
    - requests: a reference to the vector containing all the requests
    - requests_fulfilled: a reference to the atomic boolean that dictates whether the threads can stop or not
*/
fn update_account(account_to_update: &Account, requests: &Mutex<Vec<i32>>, requests_fulfilled: &AtomicBool) -> () {
    while !(requests_fulfilled.load(Ordering::Relaxed)) {
        // Lock the vector containing the requests and acquire a mutex guard
        let mut requests_guard = requests.lock().unwrap();

        if (*requests_guard).len() == 0 {
            // Checks whether the requests are all fulfilled, and if so, set the boolean to true so other threads can stop
            requests_fulfilled.store(true, Ordering::Relaxed);

            // Drop mutex guard to prevent deadlock!
            drop(requests_guard);
            return
        } else {
            // Otherwise, copy the first request (via clone) so that it persists after we remove the element from the vector!
            let current_request = (*requests_guard).get(0).unwrap().clone();

            // Erase the request from the beginning of the vector and drop the mutex guard for other threads
            (*requests_guard).remove(0);
            drop(requests_guard);

            // Add the amount in the request to the account's balance
            account_to_update.balance += current_request;
        }

    }
}

/************************************** MAIN PROGRAM **************************************/

fn main() {
    // Initialize a new account (reference counted in order for the compiler to know that it won't be dropped by any function)
    let act = Arc::new(Account { balance: Mutex::new(300) });

    // Initialize vector with all the requests to add/subtract money from account
    let requests = Arc::new(Mutex::new(make_large_vector()));
    
    // Initialize atomic boolean that will get set to true when all requests have been fulfilled (so workers can stop working)
    let all_requests_fulfilled = Arc::new(AtomicBool::new(false));

     // Initialize worker pool and pass `update_account` function into threads
    let mut workers = vec![];

    for _ in 0..6 {
        let act_ref = Arc::clone(&act);
        let requests_ref = Arc::clone(&requests);
        let req_fulfilled_ref = Arc::clone(&all_requests_fulfilled);

        let worker_thread = thread::spawn(move || update_account(&*act_ref, &*requests_ref, &*req_fulfilled_ref));
        workers.push(worker_thread);
    }

    // Wait for all threads to finish
    for worker in workers {
        worker.join().unwrap();
    }

    // Print final balance
    println!("Final balance = {:?}", *(act.balance.lock().unwrap()));
}